src/devices/Ads1115/Ads1115.cs

```diff
@@ -79,18 +78,18 @@ public class Ads1115 : IDisposable
         /// <param name="dataRate">Data Rate</param>
         /// <param name="deviceMode">Initial device mode</param>
         public Ads1115(I2cDevice i2cDevice,
-            GpioController gpioController, int gpioInterruptPin, bool shouldDispose = true, InputMultiplexer inputMultiplexer = InputMultiplexer.AIN0, MeasuringRange measuringRange = MeasuringRange.FS4096, DataRate dataRate = DataRate.SPS128,
-            DeviceMode deviceMode = DeviceMode.Continuous)
+            GpioController? gpioController, int gpioInterruptPin, bool shouldDispose = true, InputMultiplexer inputMultiplexer = InputMultiplexer.AIN0, MeasuringRange measuringRange = MeasuringRange.FS4096,
+            DataRate dataRate = DataRate.SPS128, DeviceMode deviceMode = DeviceMode.Continuous)
             : this(i2cDevice, inputMultiplexer, measuringRange, dataRate, deviceMode)
         {
-            _gpioController = gpioController ?? throw new ArgumentNullException(nameof(gpioController));
-            if (gpioInterruptPin < 0 || gpioInterruptPin >= gpioController.PinCount)
+            _gpioController = gpioController ?? new GpioController();
+            if (gpioInterruptPin < 0 || gpioInterruptPin >= _gpioController.PinCount)
             {
                 throw new ArgumentOutOfRangeException(nameof(gpioInterruptPin), $"The given GPIO Controller has no pin number {gpioInterruptPin}");
             }
 
             _gpioInterruptPin = gpioInterruptPin;
-            _shouldDispose = shouldDispose;
+            _shouldDispose = shouldDispose || gpioController is null;
```

Comment:
Wouldn't it be cleaner to have an overload that didn't take a controller, and remove the shouldDispose from there? It seems weird this parameter is only used if a value is provided. Similar example is the HttpClient that takes a handler had the dispose boolean but the handler isn't nullable - only if you use the overload that doesn't take a handler is one instantiated and shouldDispose set to true:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/HttpMessageInvoker.cs#L24-L28
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs#L130

Reply:
Good feedback. That's something we could consider, but it gets complicated really quickly because of the interplay of GPIO and I2C or SPI. Some bindings have an overload for just GPIO and another for GPIO and I2C (or SPI) where GPIO is optional. It seems like optional parameters are a perfect match for this complexity. Also, this is really the only extra line that has to exist to allow it.